### PR TITLE
[SPARK-9651] Fix UnsafeExternalSorterSuite.

### PR DESCRIPTION
**Pull request by vanzin**
First, it's probably a bad idea to call generated Scala methods
from Java. In this case, the method being called wasn't actually
"Utils.createTempDir()", but actually the method that returns the
first default argument to the actual createTempDir method, which
is just the location of java.io.tmpdir; meaning that all tests in
the class were using the same temp dir, and thus affecting each
other.

Second, spillingOccursInResponseToMemoryPressure was not writing
enough records to actually cause a spill.

### CODE REVIEW

File: core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public OutputStream apply(OutputStream stream) {
   public void setUp() {
     MockitoAnnotations.initMocks(this);
     sparkConf = new SparkConf();
-    tempDir = new File(Utils.createTempDir$default$1());
+    tempDir = Utils.createTempDir(System.getProperty("java.io.tmpdir"), "unsafe-test");
```
Comment (srowen) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: This has to be deleted in `tearDown` too I presume in order to fix this? or else it's still reusing a same dir? 

I checked and there is a similar problem in some other Java tests:

```
JavaSaveLoadSuite
JavaMetastoreDataSourcesSuite
JavaAPISuite
```

Comment (vanzin) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: In this case no; setUp and tearDown are per-test, so each test gets a separate temp dir.

I'll take a look at the other tests. Might be worth it to add the temp dir helper method in some common java class.

Comment (srowen) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: Oh right it makes a subdir every time anyway. Well, it'd be better to clean up the temp dir in `tearDown` if possible.

**Comment by JoshRosen** [comment]
Seems fine to me, pending Jenkins.

File: core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java [hunk 2]
```diff
@@ -234,7 +234,7 @@ public void testSortingEmptyArrays() throws Exception {
   public void spillingOccursInResponseToMemoryPressure() throws Exception {
     shuffleMemoryManager = new ShuffleMemoryManager(pageSizeBytes * 2);
     final UnsafeExternalSorter sorter = newSorter();
-    final int numRecords = 100000;
+    final int numRecords = (int) pageSizeBytes / 4;
```
Comment (JoshRosen) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: This didn't fail for me locally, which is kind of odd.

Comment (vanzin) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: Fails for me consistently if I use `100000`.

    Failed tests: 
    org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorterSuite.spillingOccursInResponseToMemoryPressure(org.apache.spark.util.collection.unsafe.sort.UnsafeExternalSorterSuite)
      Run 1: UnsafeExternalSorterSuite.spillingOccursInResponseToMemoryPressure:243 
    Expected: a value equal to or greater than <1>
         got: <0>
    
      Run 2: UnsafeExternalSorterSuite.tearDown:150 expected:<0> but was:<69206016>

File: core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java (hunk 1 above)
Comment (vanzin) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: Ok, I'll do this. I'll fix this test this way, then file a separate bug to have a shared, Java createTempDir() and use it from everywhere. How does that sound?

This is the only unit test that consistently fails in our builds so I'd like to at least unblock it.

Comment (vanzin) [inline core/src/test/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorterSuite.java]: Actually all the other tests you mention do the right thing. This is the only place that was calling the wrong Scala method.

**Comment by vanzin** [comment]
Well the test being changed did pass... retest this please.

**Comment by rxin** [comment]
I'm going to merge this.

**Pull request closed without merging**